tests/Microsoft.Identity.Test.Unit/CacheTests/InternalCacheOptionsTests.cs

```diff
@@ -0,0 +1,148 @@
+            Assert.AreEqual(MsalError.ExternalInternalCacheSerialization, ex.ErrorCode);
+            ex = AssertException.Throws<MsalClientException>(() => app.UserTokenCache.SetBeforeWrite((n) => { }));
+            Assert.AreEqual(MsalError.ExternalInternalCacheSerialization, ex.ErrorCode);
+
+            ex = AssertException.Throws<MsalClientException>(() => app.UserTokenCache.SetBeforeAccessAsync((n) => Task.CompletedTask));
+            Assert.AreEqual(MsalError.ExternalInternalCacheSerialization, ex.ErrorCode);
+            ex = AssertException.Throws<MsalClientException>(() => app.UserTokenCache.SetAfterAccessAsync((n) => Task.CompletedTask));
+            Assert.AreEqual(MsalError.ExternalInternalCacheSerialization, ex.ErrorCode);
+            ex = AssertException.Throws<MsalClientException>(() => app.UserTokenCache.SetBeforeWriteAsync((n) => Task.CompletedTask));
+            Assert.AreEqual(MsalError.ExternalInternalCacheSerialization, ex.ErrorCode);
+        }
+
+        [TestMethod]
+        public async Task ClientCreds_StaticCache_Async()
+        {
+            using (var httpManager = new MockHttpManager())
+            {
+                httpManager.AddInstanceDiscoveryMockHandler();
+
+                ConfidentialClientApplication app1 =
+                    ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                              .WithClientSecret(TestConstants.ClientSecret)
+                                                              .WithHttpManager(httpManager)
+                                                              .WithInternalMemoryTokenCacheOptions(new InternalMemoryTokenCacheOptions() { UseSharedCache = true }).BuildConcrete();
+
+                ConfidentialClientApplication app2 =
+                   ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                             .WithClientSecret(TestConstants.ClientSecret)
+                                                             .WithHttpManager(httpManager)
+                                                              .WithInternalMemoryTokenCacheOptions(new InternalMemoryTokenCacheOptions() { UseSharedCache = true }).BuildConcrete();
+
+                ConfidentialClientApplication app_withoutStaticCache =
+                  ConfidentialClientApplicationBuilder.Create(TestConstants.ClientId)
+                                                            .WithClientSecret(TestConstants.ClientSecret)
+                                                            .WithHttpManager(httpManager)
+                                                            .BuildConcrete();
+
+                httpManager.AddMockHandlerSuccessfulClientCredentialTokenResponseMessage();
+                await ClientCredsAssertTokenSourceAsync(app1, "S1", TokenSource.IdentityProvider).ConfigureAwait(false);
```

Comment:
This name sort of implies that you asserting the source of a token already acquired rather than doing the acquiring in the method itself. I think a more descriptive name would be
ClientCredsAcquireAndAssertTokenSourceAsync